src/datasources/workspace/WorkspaceDataSource.ts

```diff
@@ -10,30 +12,38 @@ export class WorkspaceDataSource extends DataSourceBase<WorkspaceQuery> {
     readonly templateSrv: TemplateSrv = getTemplateSrv()
   ) {
     super(instanceSettings, backendSrv);
+
+    this.variables = new WorkspaceVariableSupport();
   }
 
   baseUrl = this.instanceSettings.url + '/niuser/v1';
 
-  defaultQuery = {
-    constant: 3.14,
-  };
-
-  async runQuery(query: WorkspaceQuery, { range }: DataQueryRequest): Promise<DataFrameDTO> {
-    return {
-      refId: query.refId,
-      fields: [
-        { name: 'Time', values: [range.from.valueOf(), range.to.valueOf()], type: FieldType.time },
-        { name: 'Value', values: [query.constant, query.constant], type: FieldType.number },
-      ],
-    };
+  defaultQuery = {};
+
+  async runQuery(_query: WorkspaceQuery, { app }: DataQueryRequest): Promise<DataFrame> {
+    return toDataFrame({ fields: this.workspaceToFields(app, await this.getWorkspaces()) });
   }
 
-  shouldRunQuery(query: WorkspaceQuery): boolean {
+  shouldRunQuery(): boolean {
     return true;
   }
 
   async testDatasource(): Promise<TestingStatus> {
     await this.get(this.baseUrl + '/workspaces');
     return { status: 'success', message: 'Data source connected and authentication successful!' };
   }
+
+  private workspaceToFields(app: string, workspaces: Workspace[]): Array<{ name: string, values: string[] }> {
```

Comment:
I feel like there's a better way to write this function using the array API but my JS wheels are in need of some oil - maybe you'll know of a better implementation.